evm-ds/src/precompiles/scilla_read.rs

```diff
@@ -0,0 +1,252 @@
+        });
+    }
+    let Ok(code) = serde_json::from_slice::<Value>(&code) else {
+        return Err(PrecompileFailure::Error {
+            exit_status: ExitError::Other(Cow::Borrowed("Unable to parse scilla contract code")),
+        });
+    };
+
+    let fields = code["contract_info"]["fields"].to_owned();
+    let field_type = get_field_type_by_name(&passed_field_name, &fields)?;
+    let scilla_field = decode_indices(input, &field_type)?;
+
+    let substate_json =
+        backend.susbtate_as_json(code_address, &passed_field_name, &scilla_field.indices);
+
+    let Ok(substate_json) = serde_json::from_slice::<Value>(&substate_json) else {
+        return Err(PrecompileFailure::Error {
+            exit_status: ExitError::Other(Cow::Borrowed("Unable to parse substate json")),
+        });
+    };
+
+    let Ok(result) = extract_substate_from_json(&substate_json, &passed_field_name, &scilla_field) else {
+        return Err(PrecompileFailure::Error {
+            exit_status: ExitError::Other(Cow::Borrowed("Unable to extract return value")),
+        });
+    };
+
+    let encoded = match result {
+        Some(token) => encode(&[token]),
+        _ => Vec::default(),
+    };
+
+    Ok((
+        PrecompileOutput {
+            output_type: PrecompileOutputType::Exit(ExitSucceed::Returned),
+            output: encoded,
+        },
+        gas_needed,
+    ))
+}
+
+fn get_field_type_by_name(field_name: &str, fields: &Value) -> Result<String, PrecompileFailure> {
+    let Some(fields_array) = fields.as_array() else {
+        return Err(PrecompileFailure::Error {
+            exit_status: ExitError::Other(Cow::Borrowed("Fields in contract json are not declared as array")),
+        });
+    };
+
+    for json_field in fields_array {
+        let Some(vname) = json_field["vname"].as_str() else {
+            return Err(PrecompileFailure::Error {
+                exit_status: ExitError::Other(Cow::Borrowed("Fields in contract json are not declared as array")),
+            });
+        };
+        if vname.eq(field_name) && json_field["type"].is_string() {
+            return Ok(json_field["type"].as_str().unwrap().to_string());
+        }
+    }
+    Err(PrecompileFailure::Error {
+        exit_status: ExitError::Other(Cow::Borrowed(
+            "Unable to find field with given name in fields list",
+        )),
+    })
+}
+
+fn decode_indices(input: &[u8], input_field_type: &str) -> Result<ScillaField, PrecompileFailure> {
+    let input_field_type = input_field_type
+        .replace(['(', ')'], "")
+        .replace("Option", "");
+
+    let chunks = input_field_type.split_whitespace().collect::<Vec<_>>();
+    if chunks.len() == 1 {
+        return Ok(ScillaField {
+            indices: vec![],
+            ret_type: chunks[0].to_string(),
+        });
+    }
+    // We support only maps and maps of maps
+    let mut idx = 0;
+    let mut field_types = vec![ParamType::Address, ParamType::String];
+    while idx < chunks.len() {
```

Comment:
Worth a comment to explain what you're doing here. I don't  _think_ this has bad security implications, but I'm not entirely sure.